flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java

```diff
@@ -0,0 +1,455 @@
+	}
+
+	@Override
+	public void registerCatalog(String catalogName, Catalog catalog) {
+		catalogManager.registerCatalog(catalogName, catalog);
+	}
+
+	@Override
+	public Optional<Catalog> getCatalog(String catalogName) {
+		return catalogManager.getCatalog(catalogName);
+	}
+
+	@Override
+	public void registerFunction(String name, ScalarFunction function) {
+		functionCatalog.registerScalarFunction(
+			name,
+			function);
+	}
+
+	@Override
+	public void registerTable(String name, Table table) {
+		if (((TableImpl) table).getTableEnvironment() != this) {
+			throw new TableException(
+				"Only tables that belong to this TableEnvironment can be registered.");
+		}
+
+		CatalogBaseTable tableTable = new QueryOperationCatalogView(table.getQueryOperation());
+		registerTableInternal(name, tableTable);
+	}
+
+	@Override
+	public void registerTableSource(String name, TableSource<?> tableSource) {
+		registerTableSourceInternal(name, tableSource);
+	}
+
+	@Override
+	public void registerTableSink(
+			String name,
+			String[] fieldNames,
+			TypeInformation<?>[] fieldTypes,
+			TableSink<?> tableSink) {
+		registerTableSink(name, tableSink.configure(fieldNames, fieldTypes));
+	}
+
+	@Override
+	public void registerTableSink(String name, TableSink<?> configuredSink) {
+		// validate
+		if (configuredSink.getTableSchema().getFieldCount() == 0) {
+			throw new TableException("Table schema cannot be empty.");
+		}
+
+		checkValidTableName(name);
+		registerTableSinkInternal(name, configuredSink);
+	}
+
+	@Override
+	public Table scan(String... tablePath) {
+		return scanInternal(tablePath).map(this::createTable)
+			.orElseThrow(() -> new TableException(String.format(
+				"Table '%s' was not found.",
+				String.join(".", tablePath))));
+	}
+
+	private Optional<CatalogQueryOperation> scanInternal(String... tablePath) {
+		return catalogManager.resolveTable(tablePath)
+			.map(t -> new CatalogQueryOperation(t.getTablePath(), t.getTableSchema()));
+	}
+
+	@Override
+	public TableDescriptor connect(ConnectorDescriptor connectorDescriptor) {
+		return new StreamTableDescriptor(this, connectorDescriptor);
+	}
+
+	@Override
+	public String[] listTables() {
+		String currentCatalogName = catalogManager.getCurrentCatalog();
+		Optional<Catalog> currentCatalog = catalogManager.getCatalog(currentCatalogName);
+
+		return currentCatalog.map(catalog -> {
+			try {
+				return catalog.listTables(catalogManager.getCurrentDatabase()).toArray(new String[0]);
+			} catch (DatabaseNotExistException e) {
+				throw new ValidationException("Current database does not exist", e);
+			}
+		}).orElseThrow(() ->
+			new TableException(String.format("The current catalog %s does not exist.", currentCatalogName)));
+	}
+
+	@Override
+	public String[] listUserDefinedFunctions() {
+		return functionCatalog.getUserDefinedFunctions();
+	}
+
+	@Override
+	public String explain(Table table) {
```

Comment:
add `public String explain(boolean extended)` method to explain the whole plan for multiple-sinks ?

Reply:
We don't want to introduce new features in this PR (I know there is one exception with the lazy/eager translation, but I am actually thinking of dropping this change from this PR).